Add cheat for changing car lengths

X123M3-256 opened the pull request:
When enabled, this cheat allows users to set any train length between 1 and 64, regardless of station length or the normal maximum and minimum length of the train. It also increases the supported maximum length of trains from 16 to 64. It should be noted that rides with trains longer than 16 will not be backward compatible.

It is known not to work on tower rides or on the reverse freefall coaster - I'm not exactly sure why, but it doesn't seem too much of an issue for now. Also, if there is not room behind the station for the extra cars (either because it's a shuttle coaster or because there's a track merge), then they all just layer on top of one another. I'm not sure what can be done about it - ideally you'd limit the trains to what will fit, but that means traversing the track to determine how much room there is, which seems excessively complex.

janisozaur commented:
This will most likely break in vanilla, might depend on number of cars you select or produce undefined results. Can anyone test?

zsilencer commented:
noticed this isn't saved in network state like the other cheats

janisozaur commented:
@X123M3-256 sample commit showing how to deal with network state: https://github.com/OpenRCT2/OpenRCT2/pull/3152/commits/8d6c0c228881ee8ee1312f3f599724894e113132

Keep in mind you'll have to change `NETWORK_STREAM_VERSION` as well and thus you will clash with #3152. We can help you work that out if need be.

janisozaur commented:
Given there is a possibility of more cars per train and such large dropdowns don't really work all that well, perhaps a spinner would be a better choice?

Gymnasiast commented:
Great work! It's a cheat I have wanted to add myself for such a long time, but I could never get far enough to do it in a clean and fully working way

Broxzier commented on src/ride/ride.c:
```diff
@@ -4540,7 +4540,7 @@ train_ref vehicle_create_train(int rideIndex, int x, int y, int z, int vehicleIn
 	rct_ride *ride = get_ride(rideIndex);
 
-	uint8 trainLayout[42];
+	uint8 trainLayout[RIDE_MAX_CARS_PER_TRAIN];
```
Is the reason why this had a size of 42 known? It might be handy to add an assert here in case.
`assert(ride->num_cars_per_train <= RIDE_MAX_CARS_PER_TRAIN);`

janisozaur replied:
Such check would be useful and I intended to add such one in #3179. The problem is this code is duplicated in many places and such a check could be prone to errors, especially when refactoring it some time later on. This should really be refactored now into a small function that would make sure you can only access valid amount of cars in train.

Broxzier commented:
All rides that usually only allow for one car per train have the number of cars set to 12, as @Gymnasiast mentioned for the log flume and coaster.
The only exception I could find was the Freefall coaster, which does not allow the number of cars to be changed. No matter which one you pick, it stays at 1.

X123M3-256 commented:
@janisozaur This will definitely break in vanilla if you set more than 16 cars per train, but if you stay below that it should be backward compatible. I can't test this though, because the vanilla game doesn't work on my machine.

@zsilencer I'll see if I can fix that. I didn't really look at the network code.

@Gymnasiast You are correct that that's a problem. I think it might have something to do with the fact that single car rides normally don't have a dropdown menu - I will try to fix it.

@Broxier I have no idea why that array previously had 42 elements, but I agree that it would be a good idea to add a check for it.

Phyxion commented:
Why should this be a separate cheat? I think this could be combined with 'Unlock operating limits' (or whatever it's called).

janisozaur commented:
@Phyxion this is (at least for now) highly experimental and breaks compatibility with vanilla, I think the "unlock operating limits" is less prone to that. This should warrant enough reason for a separate option.

X123M3-256 commented:
I've noticed that those trainLayout arrays were not actually necessary; in all cases the array is created by ride_entry_get_train_layout, read once, and then discarded. I have therefore replaced calls to ride_entry_get_train_layout with a new function that just returns the entries on demand. This avoids having to allocate an array, which means the length of trains is no longer bounded by the size of that array. The limit on train length is now 255, because the train length is stored in an 8-bit integer.

However, the maximum that can be set with the dropdown has been left as 64 because the dropdown doesn't seem to work with more elements than that.

X123M3-256 commented:
I've just found this interacts badly with rides that have different colors per vehicle. Only 32 colors are allocated, which causes ride_get_vehicle color to access memory out of bounds. I don't think I can increase the size of that array without changing the save format, so I can either reduce the maximum train length to 32 or try to make it so it just uses the color of the 32nd car for all subsequent cars.

janisozaur commented:
Minor request: can you keep the oneline git commitmsg a bit shorter? It really helps when browsing the log and if you reword last commit to something like `Replace calls to ride_entry_get_train_layout with ride_entry_get_vehicle_at_position` you can still convey whole meaning and purpose. You can reword things with `git rebase -if HEAD^ ; # in editor chose the sole commit and replace "pick" with "reword" ; git push -f`

Gymnasiast commented:
@X123M3-256 The latter would be preferable. If that isn't possible, disabling 'different colors per vehicle' for rides with trains longer than 32 cars might be a better option than limiting to 32 cars.

X123M3-256 commented:
@Gymnasiast 

OK, I think I've found all the places where vehicle_colours may be accessed out of bounds, and I've made it so that it will use the the last element instead. It seems to work but I only searched src/windows/ride.c and src/ride/ride.c for potential accesses, and it doesn't rule out the possibility that there might be things other than the vehicle colours which get affected by this. 

I am still unable to figure out why it sometimes misreports the train length as 12 cars per train - the only code that looks like it might be related isn't even getting called.

@janisozaur 

I think I might screw something up trying to do that, but I'll try to keep commit messages shorter in future.

X123M3-256 commented:
I have found the cause of the misreported train lengths, but I'm not sure what I can do about it.

When you set the train length in the menu, the game sets ride->proposed_num_cars_per_train. Then ride_update_max_vehicles is called to compute the actual train length. New rides have ride-proposed_num_cars_per_train set to 12.

The problem is that ride_update_max_vehicles is called when you open the vehicle tab in the ride window. If you have a ride that has not had it's train length set manually, then ride->proposed_num_cars_per_train is 12, and when ride_update_max_vehicles is called it sets num_cars_per_train to 12 as well because there's no longer anything to limit it.

This isn't a minor problem because it actually modifies ride->num_cars_per_train. However, it does not remove and recreate the trains like would happen if you set the train length in the menu. Therefore, the num_cars_per_train no longer matches the actual length of the train. I have no idea what side-effects that might have.

Ideally, ride_update_max_vehicles would only be called if there is actually an update to make. I will have to find out where it's called from and see if this call can be removed without breaking anything.

janisozaur commented:
I got some feedback on build f0d3f1c009fe9c2353bb943c8443bf3858e4d8a3 from someone who also plays RCT:

```
-Excitement Ratings artificially high (Ratings of 11 excitement are very common,
ratings of 12 excitement are surpringly easy to achieve) 

-Reverser Rollercoaster does not enjoy long trains (breaks in indescribable ways)

-Flying coasters are not only a mess, but also maintain multiple trains, and
subsequently break on chain hills
	-Same for spinning wild mouse
		-After further testing, I have found that block breaks enable multiple
		long trains, and subsequently break them.

-Gentle rides such as Cheshire Cats allow for  multiple long trains, and don't break
	-This is actually recommended because the excitement increase works well

Observation Tower train switch allows for long length trains
	-This is how I got 640 people on Observation Tower 2 at once

Log Flumes do not operate effectively at 4 or more cars per trains
	-Same goes for Rapids and Splash Boats
	
Guest based water steering rides (Jet skis, bumper boats) cease to function with
multiple cars per train

Go-Karts also cease to function properly (Karts breaking apart, irregular
speeds, race not completing)
```

Gymnasiast commented:
@X123M3-256 I'd suggest you put a comment where the call was. Explain what the call was and why you removed it.

X123M3-256 commented:
@janisozaur I'm not sure what I can do about those issues.

The high excitement rating is because the game [applies a boost to excitement](https://github.com/OpenRCT2/OpenRCT2/blob/develop/src/ride/ride_ratings.c?#L1273) proportional to train length, and doesn't cap the maximum contribution. A cap could potentially be added, if desired.

The problem with log flumes/splash boats is because they do not use chain lifts - they are treated as powered rides, and with multiple cars the train becomes too heavy to climb the lift. It would be possible to make them more powerful but that's done by changing the .DAT file, not the code

Rides with block brakes often fail with long trains because one train can occupy two blocks. The game defaults to leaving only one block clear, so if one train occupies two blocks then all the blocks are occupied and no train can move. This can be reproduced without the cheat if you place block brakes very close together.

Flying coasters worked fine when I tested them, as did the spinning wild mouse.

Self-driven rides, tower rides, and the reverser coaster just aren't intended to have more than one car per train - I'm not surprised they glitch. Adding support is probably possible but difficult. I could try and disable the cheat on these rides if you want.

Gymnasiast commented:
I think people should expect some side effects when you use cheats like this. We warn people it isn't completely stable anyway. The block brake issue is not difficult to get when playing normally, and neither are issues with too many cars and trains - though admittedly not as severe as with this cheat.

I would only disable it on rides that break badly. Judging from the feedback janisozaur cited, those are the Go Karts, Boat Ride and Reverser Roller coaster.

X123M3-256 commented:
I think I'll just leave it as it is for now - identifying under what circumstances rides break could prove difficult (almost as hard as fixing it, to be honest), and hardcoding a list of broken rides seems inelegant (plus it would be hard to be sure that the list is exhaustive).

Gymnasiast commented:
OK, but can you put your fix for single-car vehicles being detected as 12-car ones into a commit so we can test it?

janisozaur commented:
Will you implement the spinner for car count?

X123M3-256 commented:
@Gymnasiast I've done that now. I can't think of any reason for that call to be there, but I haven't tested it that extensively.

@janisozaur I can try and do that if you want. I still need to figure out how to make it work with the network as well, and I'm not sure how to do that (or how to test it).

janisozaur commented:
@X123M3-256 regarding network:

1. serialize the cheat status in `scenario_save_network` using `SDL_WriteU8`
2. de-serialize it in `game_load_network` using `SDL_ReadU8` **in appropriate order**
3. test protocol:
  1. start server
  2. connect with a client
  3. on server disable the cheat
  4. on client try to use cheat-protected functionality, verify it fails gracefully
  5. on server enable cheat
  6. on client try to use cheat-protected functionality, verify it works

You can see it in action in 8d6c0c228881e

@IntelOrca or @zsilencer can you verify the server checks all the tries to use it (think #3140)

X123M3-256 commented:
OK, I think the network code is working. Now I need to try and work out how to replace the dropdown with a spinner.

X123M3-256 commented:
I've replaced the dropdown menu with a spinner. I have also replaced the dropdown for the number of trains with a spinner for the sake of consistency, even though this cheat doesn't affect that limit. You can now set any number of cars up to 255.

Gymnasiast commented:
I think that was the right decision.

Broxzier commented:
I'm getting these warnings now:
```
src\windows\ride.c(2381): warning C4101: 'maxCars': unreferenced local variable
src\windows\ride.c(2381): warning C4101: 'minCars': unreferenced local variable
src\windows\ride.c(2381): warning C4101: 'i': unreferenced local variable
src\windows\ride.c(2381): warning C4101: 'cars': unreferenced local variable
src\windows\ride.c(2380): warning C4101: 'stringId': unreferenced local variable
```
The position of the checkbox and label in the cheat window also looks like it's a bit off. The space before _Unlock Operating Limit_ is larger than the ones before.

Gymnasiast commented:
@Broxzier The spacing issue is a remnant from the time the description for 'Unlock operating limits' was longer. It wasn't introduced by X123M3-256.

X123M3-256 commented:
OK, I've removed those unused variables and fixed the problem with the checkbox spacing

Gymnasiast commented:
I'll test it, but I think this will get merged for 0.0.5, as we have the 0.0.4 stable coming up.

Nubbie commented on data/language/english_uk.txt:
```diff
@@ -4094,6 +4094,7 @@ STR_5786    :Invalid group name
 STR_5787    :{COMMA32} players online
 STR_5788    :Default inspection interval:
 STR_5789    :Disable lightning effect
+STR_5790    :Disable train length limit
```
This string now collides with #3219 & later on with PR #3241 
Should be changed to STR_5803

Gymnasiast replied:
Again, after 0.0.4 is released.

SceneryCoaster commented:
I have a question. What do you do to enable this? I know it's a separate version, but it seems like i have to compile it myself. (Which I have no idea how to do.) Can someone provide me with a copy of the .app?

IntelOrca commented:
@SceneryCoaster Usually builds are available for each pull request. However this branch can not currently be merged due to conflicts and the last time it was mergable, the build was unsuccessful for unrelated reasons (for Windows... Linux and OSX are fine).

X123M3-256 commented:
How do I rebase the branch? And what needs to be done to fix the merge conflicts?

janisozaur commented:
```
git fetch origin
git rebase -f origin/develop
# when conflicts arise, solve them by updating required files, save those files, then:
git add -u # to stage all tracked files
git rebase --continue
```

I think you will only get conflicts with network version and strings.

X123M3-256 commented:
Alright, I did that locally but when I try to push the changes, I get "Updates were rejected because the tip of your current branch is behind it's remote counterpart"

janisozaur commented:
Yes, that's git countermeasure for accidental removal of objects. If your branch looks good to you, just push with `-f/--force`

X123M3-256 commented:
OK, I think that has worked.

janisozaur commented:
@X123M3-256 have you tested it with saves? Do they open correctly, if you've saved it with long trains?

I understand you had not had a chance to test it with vanilla, so it still `requires testing`.

Gymnasiast commented on src/ride/ride.c:
```diff
@@ -2946,6 +2946,10 @@ track_colour ride_get_track_colour(rct_ride *ride, int colourScheme)
 {
 	vehicle_colour result;
+	//Prevent indexing array out of bounds
+		if(vehicleIndex>31) {
```
This has incorrect indentation and spacing.

X123M3-256 commented:
@Gymnasiast  Is it better now

@ Janisozaur It has been tested with saves and appears to work fine, but it has not been tested in vanilla.

Gymnasiast commented:
@X123M3-256 You forget the indentation of the closing brace. Also, I'd put a newline before the comment and after the closing brace.

X123M3-256 commented:
I've updated it again. Is it any better now?

Gymnasiast commented on src/ride/ride.c:
```diff
@@ -2946,6 +2946,12 @@ track_colour ride_get_track_colour(rct_ride *ride, int colourScheme)
+
+	//Prevent indexing array out of bounds
+	if(vehicleIndex > 31) {
+		vehicleIndex=31;
```
OpenRCT2 uses spaces around operators.

Gymnasiast commented on src/ride/ride.c:
```diff
@@ -7297,8 +7300,8 @@ static void ride_update_vehicle_colours(int rideIndex)
 			case RIDE_COLOUR_SCHEME_DIFFERENT_PER_CAR:
-				colours = ride->vehicle_colours[carIndex];
-				coloursExtended = ride->vehicle_colours_extended[carIndex];
+				colours = ride->vehicle_colours[carIndex>31?31:carIndex];
```
Spaces also needed here and on the next line.

Gymnasiast commented on src/ride/ride.c:
```diff
@@ -4543,12 +4549,9 @@ train_ref vehicle_create_train(int rideIndex, int x, int y, int z, int vehicleIn
 	train_ref train = { NULL, NULL };
 	for (int carIndex = 0; carIndex < ride->num_cars_per_train; carIndex++) {
-		rct_vehicle *car = vehicle_create_car(rideIndex, trainLayout[carIndex], carIndex, vehicleIndex, x, y, z, remainingDistance, mapElement);
+		rct_vehicle *car = vehicle_create_car(rideIndex, ride_entry_get_vehicle_at_position(ride->subtype,ride->num_cars_per_train,carIndex), carIndex, vehicleIndex, x, y, z, remainingDistance, mapElement);
```
You need to put spaces after commas.

Gymnasiast commented on src/ride/ride.c:
```diff
@@ -2946,6 +2946,12 @@ track_colour ride_get_track_colour(rct_ride *ride, int colourScheme)
 	vehicle_colour result;
+
+	//Prevent indexing array out of bounds
+	if(vehicleIndex > 31) {
```
Between `if` and `(` there should be a space.

Gymnasiast commented:
I put a few more line notes about spacing in.

Apart from that it's looking good. I hope to have time to test it once more soon. We have already slowly started merging things for 0.0.5 - not to many at the same time, in order to make testing easier.

X123M3-256 commented:
OK, those should be fixed

Gymnasiast commented on src/cheats.c:
```diff
@@ -21,6 +21,7 @@ bool gCheatsIgnoreRideIntensity = false;
 bool gCheatsDisableVandalism = false;
 bool gCheatsNeverendingMarketing = false;
 bool gCheatsFreezeClimate = false;
+bool gCheatsDisableTrainLengthLimit=false;
```
Spacing around operator.

Gymnasiast commented on src/windows/ride.c:
```diff
@@ -2449,51 +2450,19 @@ static void window_ride_vehicle_mousedown(int widgetIndex, rct_window *w, rct_wi
-
-		dropdown_set_checked(ride->num_vehicles - 1, true);
+	case WIDX_VEHICLE_TRAINS_INCREASE:
+			if(ride->num_vehicles<32)ride_set_num_vehicles(w->number, ride->num_vehicles + 1);
```
Spacing here needs adjustment.

Gymnasiast commented on src/windows/ride.c:
```diff
@@ -2449,51 +2450,19 @@ static void window_ride_vehicle_mousedown(int widgetIndex, rct_window *w, rct_wi
-
-		dropdown_set_checked(ride->num_cars_per_train - minCars, true);
+	case WIDX_VEHICLE_TRAINS_DECREASE:
+			if(ride->num_vehicles>1)ride_set_num_vehicles(w->number, ride->num_vehicles - 1);
```
Spacing here needs adjusting as well. The if is also intended a tab too far.
Both things apply to all three ifs.

Gymnasiast commented:
Some more comments. I haven't said so explicitly, but I more or less assumed you would look through the entire commit for formatting, not just the lines I pointed out. Sorry for not making that clear.

SceneryCoaster commented:
ok, post when you can have a release of the car length cheat 😉

janisozaur commented:
BTW, this would actually solve an `epic`: https://github.com/OpenRCT2/OpenRCT2/issues/714, a first one!

X123M3-256 commented:
Setting more than 32 colors per train requires changing the SV6 format, because it's a fixed 32 element array. Train length can be increased because they stored a linked lists. I've made it just use the color of the 32nd car for all cars past that as a workaround.

I have tested long trains on rides like the launched freefall (not with this cheat) - they really glitch, I don't see much use for that. It is irritating that the reverse freefall doesn't work though, I might have to see if I can fix that.

Gymnasiast commented:
I have tested it again. Some things to fix:
- The line notes I left about formatting
- The button to lower the number of cars per train is missing
- The spinner says '3 car per train', rather than '3 cars per train'

![scr187](https://cloud.githubusercontent.com/assets/1478678/14614464/999a2ffe-05a1-11e6-8913-5490bf7cad84.png)

X123M3-256 commented:
I am aware of the bug with the missing down button - it happens when the cheat is enabled while the ride window is open. I did not notice the incorrect text. However, I won't be able to fix either issue for a couple of weeks now.

janisozaur commented:
@X123M3-256 would you be fine with us making necessary changes? Your commits will be merged untouched, we would only do the cleanup we mentioned.

X123M3-256 commented:
I don't mind if you make the changes. But that bug with the down arrow not showing up needs to be fixed before this is merged - it occurs whenever you enable the cheat with the ride window still open, but once you've done that there doesn't seem to be a way to get it back without restarting the game (I really don't know what's going on).

Gymnasiast commented:
Good to hear.


A thing that stood me out: window_ride_page_enabled_widgets and window_ride_page_hold_down_widgets are very hard to understand and could do with a refactor.
